sympy/ntheory/ecm.py

```diff
@@ -0,0 +1,245 @@
+    """
+    Doubles a point in an elliptic curve in Montgomery form.
+
+    Parameter:
+    ==========
+
+    P : Point on the curve
+    a24 : Parameter for doubling
+    n : modulus
+
+    References
+    ==========
+
+    .. [1]  http://www.hyperelliptic.org/tanja/SHARCS/talks06/Gaj.pdf
+        Algorithm - 3
+    """
+
+    u, v = P.x + P.z, P.x - P.z
+    u, v = u*u, v*v
+    diff = u - v
+    x_cor = u*v % n
+    z_cor = diff*(v + a24*diff) % n
+    return Point(x_cor, z_cor)
+
+def _mont_ladder(k, P, n, a24):
+    """
+    Scalar multiblication of a point in
+    Ellipic Curve in Montgomery form.
+
+    Parameters:
+    ==========
+
+    k : The positive integer multiplier
+    P : Point on the curve in Montgomery form
+    n : modulus
+    a24 : Property of the curve
+
+    References
+    ==========
+
+    .. [1]  http://www.hyperelliptic.org/tanja/SHARCS/talks06/Gaj.pdf
+        Algorithm - 3
+    """
+
+    Q = P
+    R = _pt_double(P, n, a24)
+    for i in bin(k)[3:]:
+        if  i  == '1':
+            Q = _pt_add(R, Q, P, n)
+            R = _pt_double(R, n, a24)
+        else:
+            R = _pt_add(Q, R, P, n)
+            Q = _pt_double(Q, n, a24)
+    return Q
+
+def ecm_one_factor(n, B1=10000, B2=100000, max_curve=200):
+    """Returns one factor of n using
+    Lenstra's 2 stage Elliptic curve Factorization
+    with Suyama's Parameterization.
+
+    Parameters:
+    ===========
+
+    n : Number to be Factored
+    B1 : Stage 1 Bound
+    B2 : Stage 2 Bound
+    max_curve : Maximum number of curves generated
+
+    References
+    ==========
+
+    .. [1]  Carl Pomerance and Richard Crandall "Prime Numbers:
+        A Computational Perspective" (2nd Ed.), page 344
+    """
+
+    from sympy.ntheory import sieve, isprime
+    from sympy import gcd, mod_inverse, sqrt
+    from sympy.core.power import integer_log
+    from sympy.core.compatibility import as_int
+    import random
+
+    n = as_int(n)
+    if B1 % 2 != 0 or B2 % 2 != 0:
+        raise ValueError("The Bounds should be an even integer")
+    sieve.extend(B2)
+
+    if isprime(n):
+        return n
+
+    curve = 0
+    D = int(sqrt(B2))
+    beta = [0]*(D + 1)
+    S = [0]*(D + 1)
+    k = 1
+    for p in sieve.primerange(1, B1 + 1):
+        k *= pow(p, integer_log(B1, p)[0])
+    g = 1
+
+    while(curve <= max_curve):
+        curve += 1
+
+        #Suyama's Paramatrization
+        sigma = random.randint(6, n)
+        u = (sigma*sigma - 5) % n
+        v = (4*sigma) % n
+        diff = v - u
+        u_3 = pow(u, 3, n)
+
+        try:
+            C = (pow(diff, 3, n)*(3*u + v)*mod_inverse(4*u_3*v, n) - 2) % n
+        except ValueError:
+            #If the mod_inverse(4*u_3*v, n) doesn't exist
+            return gcd(4*u_3*v, n)
+
+        a24 = (C + 2)*mod_inverse(4, n) % n
+        Q = Point(u_3 , pow(v, 3, n))
+        Q = _mont_ladder(k, Q, n, a24)
+        g = gcd(Q.z, n)
+
+        #Stage 1 factor
+        if g != 1 and g != n:
```

Comment:
If `g == n`, then it looks like `Q.z` is zero mod `n` which means that `Q` is the neutral element (or zero) of the curve group. It cannot be used for factorization. What is done in the reference? Unfortunately, I cannot access it.